noir-projects/aztec-nr/aztec/src/state_vars/delayed_private_mutable/delayed_private_mutable_note.nr

```diff
@@ -0,0 +1,271 @@
+use dep::protocol_types::{
+    constants::{GENERATOR_INDEX__NOTE_HASH, GENERATOR_INDEX__NOTE_NULLIFIER},
+    delayed_public_mutable::DelayedPublicMutableValues,
+    hash::poseidon2_hash_with_separator,
+    traits::Packable,
+};
+
+use crate::{
+    context::PrivateContext,
+    macros::notes::MAX_NOTE_TYPES,
+    note::{
+        note_getter_options::PropertySelector,
+        note_interface::{NoteHash, NoteProperties, NoteType},
+    },
+};
+
+#[derive(Eq)]
+pub struct DelayedPrivateMutableNote<Note, let INITIAL_DELAY: u64> {
+    pub delayed_mutable_values: DelayedPublicMutableValues<Note, INITIAL_DELAY>,
+}
+
+impl<Note, let INITIAL_DELAY: u64> NoteHash for DelayedPrivateMutableNote<Note, INITIAL_DELAY>
+where
+    Note: NoteHash,
+{
+    fn compute_note_hash(self, storage_slot: Field) -> Field {
+        // There are two natural approaches we could take here:
+        // 1. Call the Note's compute_note_hash method;
+        // 2. Ignore the Note's compute_note_hash method, and instead pack the
+        //    note into fields, and hash it ourselves.
+        // I've opted for option 1, in a mad attempt to preserve "partial note"
+        // functionality.
+        //
+        // There is some redundant hashing of a storage slot within these nested
+        // compute_note_hash calls, but the opportunity to forward the
+        // responsibility is too good to pass up, for now.
+        //
+        // Let's define a partial delayed-private-mutable-note to be a _pair_ of
+        // fields:
+        //
+        // partial = [ h(storage_slot, complete(svc.pre), ...sdc.pack()), partial(svc.post) ]
```

Comment:
I haven't written any examples or tests of partial note flows, because: the pr is quite big already; there are higher priority things to do; i doubt external people would try this for a while.